boards/shared/Src/Io/Io_SharedFrequencyOnlyPwmInput.c

```diff
@@ -0,0 +1,127 @@
+ * @param frequency_hz: Frequency, in Hz
+ */
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *pwm_input,
+    const float              frequency_hz);
+
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *const pwm_input,
+    const float                    frequency_hz)
+{
+    assert(frequency_hz >= 0.0f || isnanf(frequency_hz));
+
+    pwm_input->frequency_hz = frequency_hz;
+}
+
+struct FreqOnlyPwmInput *Io_SharedFreqOnlyPwmInput_Create(
+    TIM_HandleTypeDef *htim,
+    float              timer_frequency_hz,
+    uint32_t           rising_edge_timer_channel)
+{
+    assert(htim != NULL);
+    static struct FreqOnlyPwmInput pwm_inputs[MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS];
+    static size_t                  alloc_index = 0;
+
+    assert(alloc_index < MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS);
+
+    struct FreqOnlyPwmInput *const pwm_input = &pwm_inputs[alloc_index++];
+
+    pwm_input->htim                    = htim;
+    pwm_input->timer_frequency_hz      = timer_frequency_hz;
+    pwm_input->rising_edge_tim_channel = rising_edge_timer_channel;
+    pwm_input->first_tick              = true;
+
+    HAL_TIM_IC_Start_IT(htim, rising_edge_timer_channel);
+    HAL_TIM_Base_Start_IT(htim);
+
+    return pwm_input;
+}
+
+void Io_SharedFreqOnlyPwmInput_Tick(struct FreqOnlyPwmInput *const pwm_input)
+{
+    if (pwm_input->first_tick)
+    {
+        pwm_input->curr_rising_edge = HAL_TIM_ReadCapturedValue(
+            pwm_input->htim, pwm_input->rising_edge_tim_channel);
+        pwm_input->first_tick = false;
+    }
+    else
+    {
+        pwm_input->first_tick       = true;
+        pwm_input->prev_rising_edge = pwm_input->curr_rising_edge;
+        pwm_input->curr_rising_edge = HAL_TIM_ReadCapturedValue(
+            pwm_input->htim, pwm_input->rising_edge_tim_channel);
+
+        uint32_t rising_edge_delta;
+        uint32_t prev_rising_edge = pwm_input->prev_rising_edge;
+        uint32_t curr_rising_edge = pwm_input->curr_rising_edge;
+
+        // Check if the counter value has overflowed
+        if (curr_rising_edge > prev_rising_edge)
+        {
+            rising_edge_delta = curr_rising_edge - prev_rising_edge;
+        }
+        else if (curr_rising_edge < prev_rising_edge)
+        {
+            rising_edge_delta =
+                AUTO_RELOAD_REG - prev_rising_edge + curr_rising_edge + 1;
+        }
+        else
+        {
+            rising_edge_delta = 0U;
+        }
+
+        if (rising_edge_delta != 0U)
+        {
+            Io_SetFrequency(
+                pwm_input,
+                (float)pwm_input->timer_frequency_hz / rising_edge_delta);
+        }
+        else
+        {
+            Io_SetFrequency(pwm_input, NAN);
+        }
+    }
+}
+
+size_t Io_SharedFreqOnlyPwmInput_Elapsed_Tick(
+    struct FreqOnlyPwmInput *const pwm_input,
+    size_t                         elapsed_count)
+{
+    elapsed_count++;
+
+    if (elapsed_count >= MAX_TIMER_PERIOD_ELAPSED_COUNT)
```

Comment:
I tried to understand what this function is for but I just have no clue. Can you walk me through exactly what error this is catching and how an error like that would occur?

Reply:
What I tried to do was come up with a way to detect when are are not receiving a rising edge interrupt from the flow meter. 

Each flow meter has its own elapsed count. Elapsed count for both flow meters is set to 0 at start up, and is reset to 0 when a posedge interrupt is detected on a particular channel.

When the timer counter overflows we will increment `elapsed_count`. I chose `MAX_TIMER_PERIOD_ELAPSED_COUNT =  2` and not 1 b/c there could possibly be times when we detect a rising edge interrupt right before the counter overflows (thus causing elapsed_tick to increment), followed by another rising edge interrupt right after when the timer counter restarts. But say the following rising edge interrupt doesn't arrive and the counter overflows again. This would mean we are either getting a DC signal, or trying to capture a freq that is below our measurable range (1Hz). I thought the best way to deal with this was to set the frequency as NaN.

Reply:
I mentioned this above but I changed the code for elapsed tick a little bit. I moved it into  `Io_SharedFreqOnlyPwnInput.c`

Reply:
I think this design is a bit hard to understand. I feel like we could just timestamp the last time a interrupt was received, and leave the application code to decide what to do with that information.

```
void Io_SharedFreqOnlyPwmInput_Tick()
{
  pwm_input->last_received_time = get_current_time();
  ...
}
```

@garethellis0 What do you think about this?